### PR TITLE
Nice improvements

### PR DESCRIPTION
**Pull request by johanDDC**
I fixed some inaccuracies in hessian-vector-product and added test from description.
Then I implemented such operations as TT-Matrix transpose (supports batched) and matrix-tensor-matmul, which allows apply TT-Matrix to arbitrary tensor as tensor operator.

Finally I added __neg__ and __sub__ methods to class TTBase, cause for instance in Riemann gradient descent it is not convenient to write 
`riemann_grad = (-1) * grad(x)`
instead of
`riemann_grad = -grad(x)`

### CODE REVIEW

File: ttax/autodiff_test.py [hunk 1]
```diff
@@ -44,5 +44,22 @@ def testDoubleProjection(self):
...
     self.assertAllClose(ops.full(project), ops.full(double_project), rtol=1e-4)
 
+  def testHessianVectorProduct(self):
+      rng1, rng2, rng3 = jax.random.split(jax.random.PRNGKey(0), 3)
+      dtype = jnp.float32
+      shape = (5, 5, 5)
+      A = random_.matrix(rng1, (shape, shape), dtype=dtype)
+      AT = ops.transpose(A)
+      A_plus_AT = A + AT
+      x = random_.tensor(rng2, shape, dtype=dtype)
```
Comment (Bihaqo) [inline ttax/autodiff_test.py]: The original idea was that tensors and matrices don't mix. You should only multiply matrices by matrices (but one of the matrices can be N x 1). In the case when you do want to mix them, you should explicitely convert (something like ttax.tensor_to_vector, to be implemented).

If you strongly feel otherwise, let's discuss. If not, can you please change it to
`random_.matrix(rng2, (shape, None), dtype=dtype)`

File: ttax/autodiff_test.py [hunk 2]
```diff
@@ -44,5 +44,22 @@ def testDoubleProjection(self):
...
+      vec = random_.tensor(rng3, shape, dtype=dtype)
+      proj_vec = autodiff.project(vec, x)
+
+      func = lambda x: ops.flat_inner(x, ops.tt_matvec(A, x))
```
Comment (Bihaqo) [inline ttax/autodiff_test.py]: If you do change x to a matrix, then you can just use `A @ x`

File: ttax/ops.py [hunk 3]
```diff
@@ -167,6 +167,30 @@ def flat_inner(a, b):
   return res
 
 
+def tt_matvec(A, x):
```
Comment (Bihaqo) [inline ttax/ops.py]: Again, as discussed above I would prefer to instead just have an explicit convertion between a tensor and a vector-matrix

File: ttax/ops.py [hunk 4]
```diff
@@ -368,3 +407,28 @@ def _mul_by_scalar(tt, c):
...
+    raise ValueError('The argument should be a TT-matrix.')
+
+  if tt.num_batch_dims == 0:
+    transposed_tt_cores = [
+      jnp.transpose(tt.tt_cores[core_idx], (0, 2, 1, 3)) for core_idx in range(tt.ndim)
+    ]
+  else:
+    # Batch of TT-matrices
+    transposed_tt_cores = [
+      jnp.transpose(tt.tt_cores[core_idx], (0, 1, 3, 2, 4)) for core_idx in range(tt.ndim)
```
Comment (Bihaqo) [inline ttax/ops.py]: In TTAX we support multidimensional batches (and for a good but a bit obscure reason, let me know if you  want details). So if it's a batch you don't know if it adds a single dimension or multiple.

But on the bright side, we have the tt_vmap decorator. So you can basically do this
```
@tt_vmap
def transpose(...):
  if not tt.is_tt_matrix: raise...
  transposed_tt_cores = [
      jnp.transpose(tt.tt_cores[core_idx], (0, 2, 1, 3)) for core_idx in range(tt.ndim)
    ]
    return TTMatrix(transposed_tt_cores)

File: ttax/ops_test.py [hunk 5]
```diff
@@ -212,6 +226,18 @@ def testMultiply(self):
     self.assertAllClose(res_actual1, res_desired, rtol=1e-4)
     self.assertAllClose(res_actual2, res_desired, rtol=1e-4)
 
+  def testMatrixTensorMatmul(self):
```
Comment (Bihaqo) [inline ttax/ops_test.py]: Again, remove this please.

**Review by Bihaqo — CHANGES_REQUESTED** [review]
Thank you so much for this amazing contribution, it's so nice to see this! I added a few comments, let me know if you have any questions.

File: ttax/ops.py [hunk 6]
```diff
@@ -368,3 +383,70 @@ def _mul_by_scalar(tt, c):
...
+  if not isinstance(tt, TTMatrix) or not tt.is_tt_matrix:
+    raise ValueError('The argument should be a TT-matrix.')
+
+  transposed_tt_cores = [
+    jnp.transpose(tt.tt_cores[core_idx], (0, 2, 1, 3)) for core_idx in range(tt.ndim)
+  ]
+  return TTMatrix(transposed_tt_cores)
+
+
+def vec_to_mat(tt, by_columns=True):
```
Comment (Bihaqo) [inline ttax/ops.py]: I think it's going to be easier to follow if you call it

`def tensor_to_vector(tt_tensor: TT, vertical=True) -> TTMatrix`

File: ttax/ops.py [hunk 7]
```diff
@@ -368,3 +383,70 @@ def _mul_by_scalar(tt, c):
...
+  for core in tt.tt_cores:
+    if by_columns:
+      cores.append(jnp.reshape(core, (core.shape[0], core.shape[1], 1, core.shape[2]), order="F"))
+    else:
+      cores.append(jnp.reshape(core, (core.shape[0], 1, core.shape[1], core.shape[2]), order="F"))
+  return TTMatrix(cores)
+
+
+
+def mat_to_vec(tt):
```
Comment (Bihaqo) [inline ttax/ops.py]: Similary,
`def vector_to_tensor(tt_vector: TTMatrix) -> TT`

**Review by Bihaqo — APPROVED** [review]
